# _Sprint 8-Tasca 10_ *: Hypothesis testing* 

## Level 1 
### Exercise 1

Grab a sports theme dataset you like and select an attribute from the dataset. Calculate the p-value and say if you reject the null hypothesis by taking a 5% alpha.

In [1]:
import pandas as pd
import numpy as np
import random

import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

from scipy import stats
from statsmodels.stats import weightstats as stests
from scipy.stats import f_oneway

import warnings 
warnings.filterwarnings("ignore")

import os

pd.set_option('display.max_columns', None)

In [2]:
print(os.getcwd())

/Users/jalpa/Desktop/DataFun/Course_Generalitat/Notebooks


In [3]:
os.chdir("../Data/Football/")
os.listdir()

['la-liga.csv',
 'mensfootball_results.csv',
 'laliga_player_stats_spanish.csv',
 'La_Liga_Winners.csv.xls',
 'womensfootball_results.csv',
 'laliga_player_stats_english.csv']

In [4]:
sportsdata = pd.read_csv('laliga_player_stats_english.csv')
sportsdata.head(3)

Team    Position  Shirt number          Name  Minutes played  \
0  Athletic Club  Goalkeeper           NaN  Hodei Oleaga            0.00   
1  Athletic Club  Goalkeeper           1.0     A. Remiro            0.00   
2  Athletic Club  Goalkeeper          13.0      Herrerín            2.79   

   Games played Percentage of games played  Full games played  \
0             0                       0,0%                  0   
1             0                       0,0%                  0   
2            31                      82,0%                 31   

  Percentage of full games played  Games started Percentage of games started  \
0                            0,0%              0                        0,0%   
1                            0,0%              0                        0,0%   
2                           82,0%             31                       82,0%   

   Games where substituted Percentage of games where substituted  \
0                        0                                  0,0%   
1                        0                                  0,0%   
2                        0                                  0,0%   

   Yellow Cards  Red Cards  Second Yellows  Goals scored  Penalties scored  \
0             0          0               0             0                 0   
1             0          0               0             0                 0   
2             1          0               0             0                 0   

   Own goals  Goals conceded while player on pitch  Tackles  Interceptions  \
0          0                                     0        0              0   
1          0                                     0        0              0   
2          0                                    32        0              0   

   Recoveries  Clearances  Successful tackles  Unssuccessful tackles  \
0           0           0                   0                      0   
1           0           0                   0                      0   
2         228          27                   0                      0   

   Last man  Successful duels  Duels lost  Successful aerial challenges  \
0         0                 0           0                             0   
1         0                 0           0                             0   
2         0                 3           3                            18   

   Unsuccessful aerial challenges  Offsides  Fouls suffered  Fouls committed  \
0                               0         0               0                0   
1                               0         0               0                0   
2                               1         0               4                1   

   Penalties won  Penalties given away  Handballs committed  \
0              0                     0                    0   
1              0                     0                    0   
2              0                     0                    0   

   Fouls committed per card  Shots  Shots on target  Assists  \
0                         0      0                0        0   
1                         0      0                0        0   
2                         0      0                0        1   

   Successful dribbles  Unsuccessful dribbles  Goals scored.1  \
0                    0                      0               0   
1                    0                      0               0   
2                    0                      0               0   

   From inside the area  From outside the area  Goals with left foot  \
0                     0                      0                     0   
1                     0                      0                     0   
2                     0                      0                     0   

   Goals with right foot  Penalties scored.1  Goals scored with header  \
0                      0                   0                         0   
1                      0                   0                         0   
2                      0         

In [5]:
sportsdata.shape

(556, 62)

#### Example 1

    - Attribute selected = 'Games played'
    - Null Hypothesis H_0 = Players played an average of 20 games
    - Alternative hypothesis H_1 = Players did not play an average of 20 games
    - With an alpha of 5%, if p-value < 0.05, we will reject the null hypothesis. 
    
_**The sample size is > 30 (556), therefore, I will use the Z-test for hypothesis testing.**_ 

In [6]:
attr = 'Games played'
attr_df = sportsdata[attr]
#attr_mean = np.mean(attr_df)
#print('Average Games played', attr_mean)
H_0 = 'Players played an average of 20 games' + '.'
H_1 = 'Players did not play an average of 20 games' + '.'
print("----")
ztest ,pval = stests.ztest(attr_df, x2=None, value=20)
print('p-value =',pval)
print("----")
if pval < 0.05:    # alpha value is 5% (0.05)
    print("p < 0.05: Reject null hypothesis -", H_1)
else:
    print("p > 0.05: Accept null hypothesis -", H_0)

----
p-value = 0.008189995452112976
----
p < 0.05: Reject null hypothesis - Players did not play an average of 20 games.


#### Example 2    
    
    - Attribute selected = 'Games played'
    - Null Hypothesis H_0 = Players played an average of 18 games
    - Alternative hypothesis H_1 = Players did not play an average of 18 games
    - With an alpha of 5%, if p-value < 0.05, we will reject the null hypothesis.  

In [7]:
H_0 = 'Players played an average of 18 games' + '.'
H_1 = 'Players did not play an average of 18 games' + '.'
ztest ,pval = stests.ztest(attr_df, x2=None, value=18)
print("----")
print('p-value =',pval)
print("----")
if pval < 0.05:    # alpha value is 5% (0.05)
    print("p < 0.05: Reject null hypothesis -", H_1)
else:
    print("p > 0.05: Accept null hypothesis -", H_0)

----
p-value = 0.25428347765714576
----
p > 0.05: Accept null hypothesis - Players played an average of 18 games.


## Level 2
### Exercises 2

Continue with the sports theme dataset you like and select two attributes from the dataset. Calculate the p-value and say if you reject the null hypothesis by taking a 5% alpha.

#### Example 1
    
    - Attributes selected = 'Games played', 'Games started'
    - Is there any association between players when they played games and them starting the games?
    - Null Hypothesis H_0 = There is a correlation between games played and games started
    - Alternative hypothesis H_1 = There is no correlation between games played and games started
    - With an alpha of 5%, if p-value < 0.05, we will reject the null hypothesis.


In [8]:
attr1 = 'Games played'
attr2 = 'Games started'
'''
print("Games played mean value:", np.mean(col1))
print("Games started mean value:", np.mean(col2))
print("Games played std value:", np.std(col1))
print("Games started std value:", np.std(col2))
'''
H_0 = 'There is a correlation between ' + attr1 + ' and ' + attr2 + '.'
H_1 = 'There is no correlation between ' + attr1 + ' and ' + attr2 + '.'
ztest ,pval = stests.ztest(sportsdata[attr1], sportsdata[attr2])
#ttest,pval = ttest_ind(sportsdata[attr1], sportsdata[attr2])
print("----")
print("p-value", pval)
print("----")
if pval < 0.05:
    print("p < 0.05: Reject null hypothesis -", H_1)
else:
    print("p > 0.05: Accept null hypothesis -", H_0)

----
p-value 1.988582309384357e-07
----
p < 0.05: Reject null hypothesis - There is no correlation between Games played and Games started.


#### Example 2
    
    - Attributes selected = 'Yellow Cards', 'Red Cards'
    - Is there any association between players who received yellow cards and those who received red cards?
    - Null Hypothesis H_0 = There is a correlation between yellow cards and red cards
    - Alternative hypothesis H_1 = There is no correlation between yellow cards and red cards
    - With an alpha of 5%, if p-value < 0.05, we will reject the null hypothesis.

In [9]:
attr1 = 'Yellow Cards'
attr2 = 'Red Cards'
H_0 = 'There is a correlation between players receiving ' + attr1 + ' and ' + attr2 + '.'
H_1 = 'There is no correlation between players receiving ' + attr1 + ' and ' + attr2 + '.'
ztest ,pval = stests.ztest(sportsdata[attr1], sportsdata[attr2])
print("----")
print("p-value",pval)
print("----")
if pval < 0.05:
    print("p < 0.05: Reject null hypothesis -", H_1)
else:
    print("p > 0.05: Accept null hypothesis -", H_0)

----
p-value 5.925629810494435e-109
----
p < 0.05: Reject null hypothesis - There is no correlation between players receiving Yellow Cards and Red Cards.


## Level 3
### Exercises 3

Continue with the sports theme dataset you like and select three attributes from the dataset. Calculate the p-value and say if you reject the null hypothesis by taking a 5% alpha.

#### Example 1
    
    - Attributes selected = 'Games started', 'Games played', 'Goals scored'
    - Is there any association between players who started games, played games and scoring goals?
    - Null Hypothesis H_0 = There is a correlation between Games started, Games played and Goals scored
    - Alternative hypothesis H_1 = There is no correlation between Games started, Games played and Goals scored
    - With an alpha of 5%, if p-value < 0.05, we will reject the null hypothesis.

In [15]:
attr1 = 'Games started'
attr2 = 'Games played'
attr3 = 'Goals scored'
H_0 = 'There is some correlation between one or all of these attributes: ' + attr1 + ' ' + attr2 + ' and ' + attr3 + '.'
H_1 = 'There is no correlation between one or all of these attributes: ' + attr1 + ', ' + attr2 + ' and ' + attr3 + '.'

F, pval = f_oneway(sportsdata[attr1], sportsdata[attr2], sportsdata[attr3])
#print('F=%.4f, p=%.4f' % (F, p))
print("----")
print("p-value", pval)
print("----")
if pval < 0.05:
    print("p < 0.05: Reject null hypothesis -", H_1)
else:
    print("p > 0.05: Accept null hypothesis -", H_0)

----
p-value 3.0533554796451302e-151
----
p < 0.05: Reject null hypothesis - There is no correlation between one or all of these attributes: Games started, Games played and Goals scored.


#### Example 2
    
    - Attributes selected = 'Goals scored', 'From inside the area', 'From outside the area'
    - Is there any association between scoring goals and scoring them from inside or outside the area?
    - Null Hypothesis H_0 = There is a correlation between scoring goals and scoring them from inside or outside the area
    - Alternative hypothesis H_1 = There is no correlation between scoring goals and scoring them from inside or outside the area
    - With an alpha of 5%, if p-value < 0.05, we will reject the null hypothesis.

In [11]:
attr1 = 'Goals scored'
attr2 = 'From inside the area'
attr3 = 'From outside the area'

H_0 = 'There is some correlation between one or all of these attributes: ' + attr1 + ' ' + attr2 + ' and ' + attr3 + '.'
H_1 = 'There is no correlation between one or all of these attributes: ' + attr1 + ', ' + attr2 + ' and ' + attr3 + '.'

F, pval = f_oneway(sportsdata[attr1], sportsdata[attr2], sportsdata[attr3])
print("----")
print("p-value", pval)
print("----")
if pval < 0.05:
    print("p < 0.05: Reject null hypothesis -", H_1)
else:
    print("p > 0.05: Accept null hypothesis -", H_0)

----
p-value 4.983142406364849e-26
----
p < 0.05: Reject null hypothesis - There is no correlation between one or all of these attributes: Goals scored, From inside the area and From outside the area.
